src/typechecker.erl

```diff
@@ -4734,46 +4737,101 @@ get_record_info_type({call, Anno, {atom, _, record_info},
-                        catch
-                            throw:Throw:ST ->
-                                % Useful for debugging
-                                % io:format("~p~n", [erlang:get_stacktrace()]),
-                                if
-                                    CrashOnError ->
-                                        io:format("Crashing...~n"),
-                                        erlang:raise(throw, Throw, ST);
-                                    not CrashOnError ->
-                                        [Throw | Errors]
-                                end;
-                            error:Error:ST ->
-                                %% A hack to hide the (very large) #env{} in
-                                %% error stacktraces. TODO: Add an opt for this.
-                                Trace = case ST of
-                                    [{M, F, [#env{}|Args], Pos} | RestTrace] ->
-                                        [{M, F, ['*environment excluded*'|Args], Pos} | RestTrace];
-                                    Trace0 ->
-                                        Trace0
-                                end,
-                                erlang:raise(error, Error, Trace)
-                        end;
-              (_Function, Errors) ->
-                  Errors
-          end, [], ParseData#parsedata.functions),
+    AllErrors = lists:foldr(fun (Function, Errors) ->
+                                    type_check_form_with_timeout(Function, Errors, StopOnFirstError, Env, Opts)
+                            end, [], ParseData#parsedata.functions),
     lists:reverse(AllErrors).
 
+
+%% @doc `type_check_form_with_timeout' is a workaround meant to improve usability in the
+%% presence of possible infinite loops or bugs in the typechecker.
+%%
+%% When Gradualizer is run interactively or in the foreground,
+%% detection of an infinite loop is unavoidable by the user seeing that the tool is stuck.
+%%
+%% However, when Gradualizer is run in the background (as an integration with editor / IDE)
+%% we might not realise it's stuck in an infinite loop at all or do so only after the machine's fan
+%% starts up. Moreover, in this case we'll not receive feedback from the tool.
+%%
+%% To avoid the user experience problem, it's better to opportunistically try performing the check,
+%% but in the light of it taking too long, just forcibly break the infinite loop and report
+%% a Gradualizer (NOT the checked program!) error.
+type_check_form_with_timeout(Function, Errors, StopOnFirstError, Env, Opts) ->
+    %% TODO: make FormCheckTimeOut configurable
+    FormCheckTimeOut = ?form_check_timeout_ms,
+    ?verbose(Env, "Spawning async task...~n", []),
+    TaskF = fun () ->
+                    type_check_form(Function, Errors, StopOnFirstError,
+                                    Env, Opts)
+            end,
+    Self = self(),
+    {Pid, MRef} = spawn_monitor(fun () -> Self ! TaskF() end),
```

Comment:
I suppose it's a matter of taste, so I won't demand it.

I just note that it's possible to skip one more fun here:

```suggestion
    Self = self(),
    Task = fun () ->
                   Self ! type_check_form(Function, Errors, StopOnFirstError,
                                          Env, Opts)
           end,
    {Pid, MRef} = spawn_monitor(Task)
```

Reply:
Ahh, good catch. It was necessary with the auxiliary named function hiding message passing - it's not anymore.